### DISCUSSION TITLE
Question about jax-metal development and missing operations

### DISCUSSION THREAD

**Discussion by a-gn** [jax-ml/jax#23362] [General]
Hi! I've tested jax-metal on a mac and encountered [a few of these missing operations](https://github.com/google/jax/issues?q=is%3Aopen+is%3Aissue+label%3A%22Apple+GPU+%28Metal%29+plugin%22), and others, while implementing common papers.

I'd like to know **if there is an ETA for ~full support of Apple Silicon platforms** in JAX. It would be a great help if the neural/GPU cores on my mac could replace cloud compute for some of these workflows, but right now I have to convert to numpy and back for some operations.

I'm also interested in contributing, but not sure if it's within my competence yet. Is there something I can read (part of the codebase, guide, etc.) to know where to start adding support for new operations on Metal?

Thanks!

**Comment by jakevdp** [comment]
cc/ @shuhand0 @kulinseth

**Comment by thegodone** [comment]
I see that the CI test are failing since 4 months... 
![image](https://github.com/user-attachments/assets/e57ff340-d40d-4eb1-8249-c2bde6333ebf)
Any idea why Apple plug in is not updated too ?

**Reply by shuhand0** [reply]
The failures on the CI runner are due to incompatibility between jaxlib and jax-metal. We have been working with JAX team to address it and are going to make a patch release soon.

**Comment by jakiw** [comment]
Thanks for the great library, I am using it every day.
However, some relatively important features are still missing (especially debug.print and cholesky facotrization). I have seen there have been no updates to the library since last summer, whereas it has been updated quite frequently before.

Has the library been abandoned for now // been given a very low priority? Or are updates planned?

**Reply by jakevdp** [reply]
jax-metal is closed-source, and the JAX team has no access to the codebase, so unfortunately we don't have any influence on the timeline or roadmap for jax-metal support. The Apple engineers who maintain jax-metal are tagged here (@shuhand0 & @kulinseth), so hopefully they'll be able to comment.

**Comment by clement-moulin-frier** [comment]
Hi,

I'm reviving this thread with a simple question: **should we really trust Apple to provide and maintain a usable version of JAX on Metal?**

The fact is that the [last update on Pypi is from Oct 8, 2024](https://pypi.org/project/jax-metal/), i.e. more than a year ago, and that the supposed developers never reply to any question, either here or on Apple forums (or anywhere else).

Given the importance of JAX in the ML community (and many others) and the large number of people that are willing to use JAX on their Mac, should the open-source community take action proactively instead of relying on ghost developers who might even no longer exist?

And even if Apple would make their job seriously, **do we really want that a significant part of our community depends on closed-source code managed by a company that does not care about ML research?**

I personally think this situation is very problematic, for Apple itself obviously who is loosing the trust of the ML community (but this is their problem, last time I'll buy a Mac as far as I'm concerned), but also for the whole JAX community who is loosing a large number of people who just can't use the library on their own laptop (at least not with gpu acceleration).

So should we collectively put pressure on Apple to make them release their source code, then we take over from there?

Tagging the Apple engineers in case they are suddenly willing to inform us about the status of jax-metal @shuhand0 @kulinseth
